---
title: "Liability model"
output: html_document
date: "2025-06-17"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
N = 10000 # of individuals
K = 0.1 # Ks <- c(0.01,0.05,0.1,0.2), lifetime prevalence of disease
M = 1000 # of SNPs
h2 = 0.5 # heritability TODO: vary heritability
```

Liability model in Visscher et al.

```{r}
U  <- seq(1/N,1-1/N,length=N) # sequence of probabilities between 0 and 1
x  <- qnorm(U) # convert the probabilities to quantiles of the standard normal distribution (liability scores)
t  <- qnorm(1-K) # threshold liability for disease based on prevalence K. People with liability > t have the disease
R0 <- pnorm(x,mean=-t,sd=1,lower.tail = F) # risk distribution in the population, risk at each liability point x
z = dnorm(t) # density of standard normal distribution at t

```

-   Generate random genotypes
-   Somehow compute the liability for each individual, and a phenotype based on the liability
-   Compute PGSs for each individual
    -   this would require both a training and a validation set before getting to evaluate the performance

```{r}
ps = runif(M, 0, 1) #  allele 1 frequencies for M SNPs
var = h2/((2*ps*(1-ps))^S*M) # variance proportional to negative selection coefficient, scaled to match heritability h2 and M SNPs
betas = rnorm(M, mean = 0, sd = sqrt(var)) # draw betas with variance var.neg
betas.scaled = betas*K*(1-K)/z
beta.est = rnorm(M, betas.scaled, sqrt(1/N *(1 - var))) # add noise proportional to sample size, formula from paper 3

X <- matrix(rbinom(N * M, size = 2, prob = rep(ps, each = N)), nrow = N, ncol = M) # generate genotypes assuming allele frequencies follow HWE
```

-   generate genetic effects G
-   generate environmental effects E
-   liability L = G + E

#### Random genetic effects

```{r}
G = rnorm(N, 0, sqrt(h2)) # genetic effects
E = rnorm(N, 0, sqrt(1-h2)) # environmental effects
L = G + E # liability values
length(which(L > t))
disease.status = ifelse(L > t, 1, 0) # assign disease status
```

```{r}
cat("Simulated disease prevalence:", mean(disease.status), "\n")

# Plot liability distribution
library(ggplot2)
df <- data.frame(Liability = L, Disease = factor(disease.status))

ggplot(df, aes(x = Liability, fill = Disease)) +
  geom_density(alpha = 0.5) +
  geom_vline(xintercept = t) +
  labs(title = "Liability threshold model",
       subtitle = paste("simulated disease prevalence =", round(mean(disease.status),3)),
       fill = "Disease Status") +
  theme_minimal()

```

#### Genetic effects proportional to PGS

Generate genotypes and compute PGS using PGS.uncertainty

```{r}
PGSs = PGS.uncertainty(N, M, h2, -1) # Generate genotypes and compute PGS
PGS = PGSs$pgs
PGS.scaled = scale(PGS) # scale PGSs to have mean = 0 and var = 1
e = rnorm(N, 0, sqrt(1-h2)) # environmental effects
g = PGS.scaled * sqrt(h2) # genetic effects 
l = g + e # liability values

length(which(l > t))
disease.status = ifelse(l > t, 1, 0) # assign disease status
```

```{r}
cat("Simulated disease prevalence:", mean(disease.status), "\n")

# Plot liability distribution
library(ggplot2)
df <- data.frame(Liability = l, Disease = factor(disease.status))

ggplot(df, aes(x = Liability, fill = Disease)) +
  geom_density(alpha = 0.5) +
  geom_vline(xintercept = t) +
  labs(title = "Liability threshold model",
       subtitle = paste("simulated disease prevalence =", round(mean(disease.status),3)),
       fill = "Disease Status") +
  theme_minimal()
```

#### Effect sizes transformed to the liability scale

Generate genotypes and compute PGS using PGS.uncertainty.liability

```{r}
source("PGSvariance.R")
PGSs = PGS.uncertainty.liability(N, M, h2, -1, K) # Generate genotypes and compute PGS
PGS = PGSs$pgs
PGS.scaled = scale(PGS) # scale PGSs to have mean = 0 and var = 1
e = rnorm(N, 0, sqrt(1-h2)) # environmental effects
g = PGS.scaled * sqrt(h2) # genetic effects

l = g + e # liability values

length(which(l > t))
disease.status = ifelse(l > t, 1, 0) # assign disease status
```

```{r}
cat("Simulated disease prevalence:", mean(disease.status), "\n")

# Plot liability distribution
library(ggplot2)
df <- data.frame(Liability = l, Disease = factor(disease.status))

ggplot(df, aes(x = Liability, fill = Disease)) +
  geom_density(alpha = 0.5) +
  geom_vline(xintercept = t) +
  labs(title = "Liability threshold model",
       subtitle = paste("simulated disease prevalence =", round(mean(disease.status),3)),
       fill = "Disease Status") +
  theme_minimal()
```
#### Constant environmental effects with varying genetic effects

With constant environmental effects for each individual, we sample genetic effect estimates and compute the genetic component using PGS for each sample. Look at which proportion of the liability values exceeds the threshold for each individual. 

```{r}
source("PGSvariance.R")
N = 10000 # of individuals
K = 0.1 # Ks <- c(0.01,0.05,0.1,0.2), lifetime prevalence of disease
M = 1000 # of SNPs
h2 = 0.5 # heritability 
n.samples = 1000

# sample environmental effects
e = rnorm(N, 0, sqrt(1-h2))

# sample PGSs for each individual
pgs.s = PGS.sample.uncertainty.liability(N, M, h2, S = -1, K, n.samples = 1000)
pgs.samples = pgs.s$pgs.samples # a (n.samples x N) matrix: each column is a sample of PGSs for an individual

# compute genetic effects for each pgs sample
pgs.scaled = scale(t(pgs.samples)) # scale each row to have mean = 0 and var = 1, (N x n.samples) matrix
g = pgs.scaled * sqrt(h2) # genetic effects, (N x n.samples) matrix of n.samples genetic effects for each individual

# calculate liabilities, n.samples liabilities for each individual
l = e + g # a (N x n.samples) matrix of b.samples liabilities for each individual

t = qnorm(1-K) # threshold liability for disease based on prevalence K. People with liability > t have the disease
over.t = 1 * (l > t) # a (N x n.samples) matrix of boolean values, indicating whether the liability value is over the threshold
prop.over.t = rowSums(over.t)/n.samples # proportion of liability samples over threshold

prop.over.t[order(prop.over.t, decreasing = TRUE)]

sum( ( prop.over.t > 0.4 & prop.over.t < 0.6) * 1) # of individuals with proportion of liability values over threshold in [0.4, 0.6]

```

with N = 10 000, 440 individuals with proportion of liability values over threshold in [0.4, 0.6]
with N = 100 000, 4240 individuals with proportion of liability values over threshold in [0.4, 0.6]
with N = 300 000, 11 009 individuals with proportion of liability values over threshold in [0.4, 0.6]

-> compute proportion of individuals having proportion of samples over t in (0.4, 0.6) for different N

#### Constant genetic effects with varying environmental effects

```{r}
source("PGSvariance.R")
N = 10000 # of individuals
K = 0.1 # Ks <- c(0.01,0.05,0.1,0.2), lifetime prevalence of disease
M = 1000 # of SNPs
h2 = 0.5 # heritability 
n.samples = 1000

# sample N different genetic effects
PGSs = PGS.uncertainty.liability(N, M, h2, -1, K) # Generate genotypes and compute PGS
PGS = PGSs$pgs
PGS.scaled = scale(PGS) # scale PGSs to have mean = 0 and var = 1
g = as.vector(PGS.scaled * sqrt(h2)) # genetic effects

# sample environmental effects
e = matrix(rnorm(N*n.samples, 0, sqrt(1-h2)), nrow = N, ncol = n.samples)

# calculate liabilities, n.samples liabilities for each individual
l = e + g # a (N x n.samples) matrix of n.samples liabilities for each individual

t = qnorm(1-K) # threshold liability for disease based on prevalence K. People with liability > t have the disease
over.t = 1 * (l > t) # a (N x n.samples) matrix of boolean values, indicating whether the liability value is over the threshold
prop.over.t = rowSums(over.t)/n.samples # proportion of liability samples over threshold: "risk of disease with this genotype"
prop.over.t

#prop.over.t[order(prop.over.t, decreasing = TRUE)]

#sum( ( prop.over.t > 0.4 & prop.over.t < 0.6) * 1) # of individuals with proportion of liability values over threshold in [0.4, 0.6]


```

#### Liability of individuals in top PGS quantile
cmd + option + i


```{r}
# Liability of individuals in the top PGS quantile
N = 10000 # of individuals
K = 0.01 # Ks <- c(0.01,0.05,0.1,0.2), lifetime prevalence of disease
M = 1000 # of SNPs
h2 = 0.8 # heritability 
tPGS = 0.9

# generate genotypes and effect sizes for individuals
PGSs = PGS.uncertainty.liability(N, M, h2, -1, K) # Generate genotypes and compute PGS
PGS = PGSs$pgs 
g = as.vector(scale(PGS) * sqrt(h2)) # scaled genetic effects

ranks = which(N*tPGS < rank(PGS, ties.method = "first")) # individual PGSs in the top PGS quantile

e = rnorm(N, 0, sqrt(1-h2)) # environmental effects

l = g + e # liability values for each individual
hd = which(l > t) # individuals above threshold t

# compute proportion of individuals in top PGS quantile that have disease (exceed the liability threshold)
length(intersect(ranks, hd))/length(ranks) # proportion of top quantile PGS individuals that have the disease
```





